diff --git a/src/libstd/sys/unix/mod.rs b/src/libstd/sys/unix/mod.rs
index bbed42c..388d2fc 100644
--- a/src/libstd/sys/unix/mod.rs
+++ b/src/libstd/sys/unix/mod.rs
@@ -27,7 +27,21 @@ use ops::Neg;
 #[cfg(target_os = "netbsd")]    pub use os::netbsd as platform;
 #[cfg(target_os = "openbsd")]   pub use os::openbsd as platform;
 
+#[cfg(feature = "backtrace")]
 pub mod backtrace;
+
+#[cfg(not(feature = "backtrace"))]
+pub mod backtrace {
+    use prelude::v1::*;
+    use io::prelude::*;
+
+    use io;
+
+    pub fn write(_: &mut Write) -> io::Result<()> {
+        Ok(())
+    }
+}
+
 pub mod c;
 pub mod condvar;
 pub mod ext;
